api4/work_templates.go

```diff
@@ -0,0 +1,146 @@
+			http.NotFound(w, r)
+			return
+		}
+		if !hasBoard {
+			c.App.Log().Warn("trying to access work templates api while while board product is not installed")
+			http.NotFound(w, r)
+			return
+		}
+
+		h(c, w, r)
+	}
+}
+
+func getWorkTemplateCategories(c *Context, w http.ResponseWriter, r *http.Request) {
+	t := c.AppContext.GetT()
+
+	categories, appErr := c.App.GetWorkTemplateCategories(t)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	b, err := json.Marshal(categories)
+	if err != nil {
+		c.Err = model.NewAppError("getWorkTemplateCategories", "api.marshal_error", nil, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(b)
+}
+
+func getWorkTemplates(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireCategory()
+	if c.Err != nil {
+		return
+	}
+	t := c.AppContext.GetT()
+
+	workTemplates, appErr := c.App.GetWorkTemplates(c.Params.Category, c.App.Config().FeatureFlags.ToMap(), t)
+	if appErr != nil {
+		c.Err = appErr
+		return
+	}
+
+	b, err := json.Marshal(workTemplates)
+	if err != nil {
+		c.Err = model.NewAppError("getWorkTemplates", "api.marshal_error", nil, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(b)
+}
+
+func executeWorkTemplate(c *Context, w http.ResponseWriter, r *http.Request) {
+	wtcr := &worktemplates.ExecutionRequest{}
+	err := json.NewDecoder(r.Body).Decode(wtcr)
+	if err != nil {
+		c.Err = model.NewAppError("executeWorkTemplate", "api.unmarshal_error", nil, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	canCreatePublicChannel := c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), wtcr.TeamID, model.PermissionCreatePublicChannel)
+	canCreatePrivateChannel := c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), wtcr.TeamID, model.PermissionCreatePrivateChannel)
+	// focalboard uses channel permissions for board creation
+	canCreatePublicBoard := canCreatePublicChannel
+	canCreatePrivateBoard := canCreatePrivateChannel
+	canCreatePublicPlaybook := c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), wtcr.TeamID, model.PermissionPublicPlaybookCreate)
+	canCreatePrivatePlaybook := c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), wtcr.TeamID, model.PermissionPrivatePlaybookCreate)
+	canCreateRun := c.App.SessionHasPermissionToTeam(*c.AppContext.Session(), wtcr.TeamID, model.PermissionRunCreate)
+	canExecuteWorkTemplate, err := wtcr.CanBeExecuted(worktemplates.PermissionSet{
+		CanCreatePublicChannel:   canCreatePublicChannel,
+		CanCreatePrivateChannel:  canCreatePrivateChannel,
+		CanCreatePublicBoard:     canCreatePublicBoard,
+		CanCreatePrivateBoard:    canCreatePrivateBoard,
+		CanCreatePublicPlaybook:  canCreatePublicPlaybook,
+		CanCreatePrivatePlaybook: canCreatePrivatePlaybook,
+		CanCreatePlaybookRun:     canCreateRun,
+	})
+	if err != nil {
+		if canExecuteWorkTemplate == nil {
+			c.Err = model.NewAppError("executeWorkTemplate", "api.execute_work_template_error", nil, err.Error(), http.StatusForbidden)
```

Comment:
The idiomatic way to construct AppError is to use the wrap method: `model.NewAppError("executeWorkTemplate", "api.execute_work_template_error", nil, "", http.StatusForbidden).Wrap(err)`